### ISSUE TITLE
stripping `\t` from urls

### ISSUE THREAD

**Issue by iscai-msft** [aio-libs/aiohttp#5904]
### Describe the bug

aiohttp is stripping `\t` from our tsv formatted URLs. For example, when I give aiohttp a url in the following format `http://localhost:3000/...\tbegin...end\t\t`, when it [creates a URL object](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L366) out of my string, the `\t` portions of the URL are stripped, and this information is no longer there when it makes it's way to the server

### To Reproduce

1. Send a request to a url with `\t` in it, and during the call to `async def _request` when it creates a URL object, the `\t` is stripped out

### Expected behavior

I would expect the `\t`s to stay in the URL, this is the behavior I'm getting from requests

### Logs/tracebacks

```python-traceback
I'm debugging, but the code that's getting rid of the `\t` is the code that creates the URL object https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L366
```

### Python Version

```console
$ python --version
3.9.5
```

### aiohttp Version

```console
$ python -m pip show aiohttp
3.7.4.post0
```

### multidict Version

```console
$ python -m pip show multidict
5.1.0
```

### yarl Version

```console
$ python -m pip show yarl
1.6.3
```

### OS

macOS

### Related component

Client

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by Dreamsorcerer** [comment]
Well, tabs are not valid characters in a URL, so it percent-encodes them (I don't see it stripping them):
```
>>> yarl.URL("http://my.domain/\t")
URL('http://my.domain/%09')
```

**Comment by iscai-msft** [comment]
thank you @Dreamsorcerer , I'm not able to repro your repro though, here's what I get
<img width="260" alt="Screen Shot 2021-07-20 at 3 56 46 PM" src="https://user-images.githubusercontent.com/43154838/126387455-463a60cf-0585-43f0-a704-fdcd30b8b3f2.png">

I wonder if this is because of a difference in version of multidict / yarl, my versions are `1.6.3` for yarl and `5.1.0` for multidict (I believe the latest for both)

**Comment by Dreamsorcerer** [comment]
Hmm, that's weird.
```
>>> import yarl                        
>>> yarl.URL("http://my.domain/\t\t\t")
URL('http://my.domain/%09%09%09')
>>> yarl.__version__
'1.6.3'
>>> import multidict
>>> multidict.__version__
'5.1.0'
```

I can see there are 2 quoting implementations, try the pure Python one and see if the problem still occurs:
```
>>> import os
>>> os.environ["YARL_NO_EXTENSIONS"] = "1" 
>>> import yarl                            
>>> yarl._quoting._Quoter  # Check this is _quoting_py, not _quoting_c.
<class 'yarl._quoting_py._Quoter'>
>>> yarl.URL("http://my.domain/\t\t\t")    
URL('http://my.domain/%09%09%09')
```

**Comment by iscai-msft** [comment]
Thank you so much!

I tried the pure Python quoting implementation, I'm still not seeing %09
<img width="345" alt="Screen Shot 2021-07-20 at 5 02 29 PM" src="https://user-images.githubusercontent.com/43154838/126395089-6efc35bb-4b8c-40fe-a9be-2abc62420651.png">

To add a bit more context, I'm using Mac OS 11.4 and it appears that the yarl behavior is correct in some OSs and not in others. Using ubuntu image `ubuntu-16.04` my `\t` is being correctly encoded and kept. However, using my Mac OS and `ubuntu-20.04`, I'm running into the issue where the `\t` is being dropped

**Comment by Dreamsorcerer** [comment]
I'm pretty much out of ideas...
Just to reduce it down to the minimum, try out this:
```
>>> from yarl._quoting_py import _Quoter as _PyQuoter
>>> from yarl._quoting_c import _Quoter as _CQuoter
>>> _PyQuoter()("\t")
'%09'
>>> _CQuoter()("\t") 
'%09'
```

I ran a test using the CI at https://github.com/aio-libs/yarl/pull/606
That works as expected, running on Windows, Mac, Linux, with Python 3.6-3.9.

Without some way to reproduce the issue, I'm not sure what else I can do.

**Comment by iscai-msft** [comment]
Just want to say thank you again @Dreamsorcerer . The code I copied from you is working the same as yours, but the final result is still without the '%09'

<img width="399" alt="Screen Shot 2021-07-20 at 5 16 23 PM" src="https://user-images.githubusercontent.com/43154838/126396510-31e2ec38-9532-4ca5-bb9c-170031c08220.png">

I went ahead and tested the test you added (thanks so much again!) and it was _not_ able to repro my bug, so I took the liberty of creating a test that repro by bug. Was going to push but got denied permission, so probably easier to just copy paste here. Can you give this a try and see if it repros the issue?

```python
def test_tab():
    url = URL("http://my.domain/\t\t\t")
    assert str(url) == 'http://my.domain/%09%09%09'
```

**Comment by Dreamsorcerer** [comment]
Interesting, so there must be something happening somewhere between `URL` and the quoting.

Weird, it succeeds on Windows and Pypy, but is failing on Ubuntu and Mac, despite the fact it works successfully on my laptop also running Ubuntu...
https://github.com/aio-libs/yarl/pull/607

**Comment by iscai-msft** [comment]
ok yay, glad it repros now! Yeah I noticed when running in the pipelines, that ubuntu 16.4 is working, but 20.4 is failing, so somewhere on the ride it stopped working for ubuntu

**Comment by Dreamsorcerer** [comment]
Hmm, well, I'm on Ubuntu 20.04 and it's working on both the host and in an lxc container.

**Comment by Hanaasagi** [comment]
It works on
- Linux 5.10.0-7-amd64 SMP Debian 5.10.40-1 (2021-05-28) x86_64 GNU/Linux.
- yarl Version: 1.6.3


Could you test this code example? It simulates the logic of `yarl` processing a `\t` character. 

```Python
from yarl import URL
from urllib.parse import urlsplit


url = "http://my.domain/\t"
print("Test url: ", repr(url))
val = urlsplit(url)
print("Use Quoter: ", URL._PATH_REQUOTER.__dict__)
print(("Manual encode: %{:02X}".format("\t".encode('utf8')[0])).encode("ascii"))
print("Before: ", repr(val[2]))
path = URL._PATH_REQUOTER(val[2])  # type: ignore
print("After: ", repr(path))

assert path == "/%09"

```

**Comment by iscai-msft** [comment]
Thanks for the code snippet @Hanaasagi . The logic is failing for me (which I guess is good news?)

For more information, I'm using:
1. Python 3.9.1
2. macOS-10.16-x86_64-i386-64bit
3. yarl `1.6.3`

<img width="607" alt="Screen Shot 2021-07-21 at 3 00 32 PM" src="https://user-images.githubusercontent.com/43154838/126544813-6bfed9c5-187b-4af1-a4be-b7ec8239f62f.png">

**Comment by Hanaasagi** [comment]
Thanks for your reply. `urlsplit` strips the `\t`. The recent Python version changed this behavior.

From https://docs.python.org/3.8/library/urllib.parse.html#urllib.parse.urlsplit

> Following the WHATWG spec that updates RFC 3986, ASCII newline \n, \r and tab \t characters are stripped from the URL.

If you are using following version (or above), `\t` will be stripped.

- 3.7.11
- 3.8.10
- 3.9.5
- 3.10

More info, please see https://bugs.python.org/issue43882.

**Comment by Dreamsorcerer** [comment]
Seems like it was only partially patched in 3.8.10, which might be why I'm not seeing it.

But, seems like this is the correct behaviour and we should expect tabs and newlines to not work in URLs.
So, repeating my initial recommendation, I'd recommend moving this structured data to the body of a request, if possible. Sending this kind of data in a URL is bad practice and going to fail in many situations (not just aiohttp).
